dlib/te.h

```diff
@@ -0,0 +1,440 @@
+// Copyright (C) 2022  Davis E. King (davis@dlib.net)
+// License: Boost Software License   See LICENSE.txt for the full license.
+#ifndef DLIB_TYPE_ERASURE_H_
+#define DLIB_TYPE_ERASURE_H_
+
+#include <type_traits>
+#include <utility>
+#include <typeindex>
+
+namespace dlib
+{
+    namespace te
+    {
+        class bad_te_cast : public std::bad_cast 
+        {
+        public:
+            virtual const char * what() const throw()
+            {
+                return "bad_te_cast";
+            }
+        };
+
+        struct storage_heap
+        {
+            storage_heap() = default;
+
+            template <
+                class T,
+                class T_ = std::decay_t<T>,
+                std::enable_if_t<!std::is_same<T_,storage_heap>::value, bool> = true
+            >
+            storage_heap(T &&t) noexcept(std::is_nothrow_constructible<T_,T&&>::value)
+            :   ptr{new T_{std::forward<T>(t)}},
+                del{[](void *self) { 
+                    delete reinterpret_cast<T_*>(self); 
+                }},
+                copy{[](const void *self) -> void * {
+                    return new T_{*reinterpret_cast<const T_*>(self)};
+                }},
+                type_id{[] {
+                    return std::type_index{typeid(T_)};
+                }}
+            {
+            }
+
+            template <
```

Comment:
I would not bother defining this. Whatever uses a storage object should implement these types of constructors themselves.